core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -268,8 +274,10 @@ class ReplicaManager(val config: KafkaConfig,
   def getOrCreatePartition(topic: String, partitionId: Int): Partition = {
     var partition = allPartitions.get((topic, partitionId))
     if (partition == null) {
-      allPartitions.putIfNotExists((topic, partitionId), new Partition(topic, partitionId, time, this))
-      partition = allPartitions.get((topic, partitionId))
+      val newPartition = new Partition(topic, partitionId, time, this)
+      partition = allPartitions.putIfNotExists((topic, partitionId), newPartition)
+      if (partition == null)
+          partition = newPartition
     }
     partition
```

Comment:
@ijuma I imagine the `get` was used for lazy create of `Partition`. I have rewritten this using `Pool.getAndMaybePut`.